### ISSUE TITLE
System.IO.FileNotFoundException -> Loading of assemblies failure

### ISSUE THREAD

**Issue by fromberg100** [dotnet/aspnetcore#48508]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When I run my project on my local environment using "dotnet run" everything works fine.

However, when I publish my project (self-contained) and after I run the .exe I start getting exceptions for certain assemblies:

System.IO.Pipes
System.Security.AccessControl
System.IO.Compression
System.Drawing

For all of them the exception is the same, example:
**System.IO.FileNotFoundException: Could not load file or assembly 'System.IO.Pipes, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'**

All dlls are in the publish folder

As workaround, I do the following on Program.cs:
var pipesAssembly = typeof(System.IO.Pipes.PipeStream).Assembly;
var accessControlAssembly = typeof(System.Security.AccessControl.AccessRule).Assembly;
var compressionAssembly = typeof(System.IO.Compression.ZipArchive).Assembly;
var drawingAssembly = typeof(System.Drawing.Bitmap).Assembly;

If I do the above, the project runs fine, but I am concerned having to do this.

### .NET Version

7.0.302

### Anything else?

I am attaching the .deps.json file generated after publishing my project.
[deps.json.zip](https://github.com/dotnet/aspnetcore/files/11593222/deps.json.zip)

**Comment by davidfowl** [comment]
What sort of application is it?

**Comment by fromberg100** [comment]
Web Application (ASP.NET Core 7.0.5)

**Comment by davidfowl** [comment]
Are you trimming the application?

**Comment by fromberg100** [comment]
no, I am not trimming the application.

Please note I did a recent upgrade from .NET 5 to .NET 7.  With .NET 5 I did not have the issues I reported, I was not trimming with .NET5 either.

**Comment by halter73** [comment]
What's loading System.IO.Pipes, System.Drawing.Bitmap, etc..? Is it some sort of plugin loading logic that's thrown off by all the assemblies combined into the single-file executable for the self-contained publish?

**Comment by fromberg100** [comment]
Exceptions are thrown depending on what controller's action is called.

For example, if I upload an image, the controller's action creates an instance of System.Drawing.Bitmap to create different sizes of the original image.  That's when I get the exception Could not load file or assembly 'System.Drawing, Version=7.0.0.0

Another example, I use DocumentFormat.OpenXml to read and write xlsx files.  When the controller's action writes a xlsx file I get the exception Could not load file or assembly 'System.IO.Compression, Version=7.0.0.0

When I publish, assemblies are not combined into a single file.  All dlls are there in the publish folder:

![image](https://github.com/dotnet/aspnetcore/assets/13880898/2f396e92-6e27-4d5d-b0b5-ed3c49507509)

![image](https://github.com/dotnet/aspnetcore/assets/13880898/770ce2a8-5edb-4abb-a63c-3a8ecf4955b6)

**Comment by fromberg100** [comment]
I will create a separate/simple web application project tonight and will try to reproduce the System.Drawing exception and will post it here.

**Comment by fromberg100** [comment]
I am still working on this.  I noticed something.

If I add the package reference Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation 7.0.5, when I publish my web application, 
the deps.json file contains a lot of dependencies:


![image](https://github.com/dotnet/aspnetcore/assets/13880898/1b83d4a5-0713-4bd2-8e2a-b481590f8222)

it is a long list, please see the file attached to see the whole thing:
[WebApplication1.deps.zip](https://github.com/dotnet/aspnetcore/files/11618497/WebApplication1.deps.zip)

As you can see it adds System.Drawing 4.0.0:

![image](https://github.com/dotnet/aspnetcore/assets/13880898/ae266339-7e2f-4238-944d-bd60b0d2a8d4)

If I remove the package reference Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation, the dependencies are just a few:

![image](https://github.com/dotnet/aspnetcore/assets/13880898/f8f83f86-7e45-4932-b004-76cb6ed47bfd)

**Comment by fromberg100** [comment]
Sorry for not providing feedback before.

Unfortunately I could not create a simple project to reproduce the issue  and share it here.

My project uses many other referenced projects and external libraries.  I believe it has to do with one in particular.  My workaround is to load all assemblies upon project startup, see below:

`var rootAssembly = Assembly.GetEntryAssembly();

var visited = new HashSet<string>();
var queue = new Queue<Assembly>();

queue.Enqueue(rootAssembly);

while (queue.Any())
{
	var assembly = queue.Dequeue();
	visited.Add(assembly.FullName);

	var references = assembly.GetReferencedAssemblies();
	foreach (var reference in references)
	{
		if (!visited.Contains(reference.FullName))
		{
			var loadedAssembly = Assembly.Load(reference);
			queue.Enqueue(loadedAssembly);

			visited.Add(reference.FullName);
		}
	}
}`

I am willing to share my screen if needed so you can see the issue I am having.

**Comment by GOATS2K** [comment]
This is still an issue. I am unable to reproduce it in a smaller project - but mine is a simple console application with Spectre.Console and RestSharp - my application crashes when I call `Process.Start("random-process)`.

This happens on .NET 8 RC 1 and any version before that.

**Comment by AndreiMaz** [comment]
We also experience the same issue in nopCommerce. Please find more information about it at https://www.nopcommerce.com/en/boards/active-discussions

**Comment by batkaevruslan** [comment]
> We also experience the same issue in nopCommerce. Please find more information about it at [nopcommerce.com/en/boards/active-discussions](https://www.nopcommerce.com/en/boards/active-discussions)

@AndreiMaz , hi! I've read the whole thread https://www.nopcommerce.com/en/boards/topic/97606/460-html-formatting-keeps-disappearing-/page/13 and as I understand you have not found the cause of the problem?



Below I'll share some thoughts from the side of our project.

We have problems with loading:
* System.Net.WebHeaderCollection
* System.IO.Compression.ZipFile
* System.Xml.XPath.XDocument
* System.Net.Requests

We have been using sdk 7.302 (the same as in the issue description) for at least 3 months, but the errors appeared first just 2 weeks ago. 
The errors appear on different environments randomly: we are not able to reproduce it, just see them in logs.
All our dependencies are up-to-date thanks to Dependabot: it is one of possible reasons for the errors.
Recently we've:
* specified RuntimeIdentifier for our web-projects to reduce the size of `bin` folder
* removed NuGet reference to `Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation` from one of our library projects (it was accidently added several months ago). 
* disabled "most" of SatelliteResourceLanguages to reduce `bin` folder size

**Comment by AndreiMaz** [comment]
@batkaevruslan Actually we've found the solution (currently testing). Please see [this post](https://www.nopcommerce.com/en/boards/topic/97606/460-html-formatting-keeps-disappearing-/page/12#308559). Could you please also test it?

**Comment by batkaevruslan** [comment]
@AndreiMaz, thanks for sharing! We'll try your solution if we won't be able to find the root cause. In any case I'll share the results.

**Comment by batkaevruslan** [comment]
@fromberg100 , @GOATS2K , @AndreiMaz , we ran [Process Monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) on the server with our web app and noticed that the application tries to load "problem" dll `System.Xml.XPath.XDocument, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a` from path **without** "dll" extensions:
![image](https://github.com/dotnet/aspnetcore/assets/3961279/541e7197-932c-45d4-b568-51bd319e5a28)

Could you use Process Monitor and check if you have the same behavior?

We are trying to reproduce it locally. The best of what we've got by this time: 
* we manually removed System.Xml.XPath.XDocument.dll from `bin` folder
* ran application
* saw in debug that AssemblyLoadContext tries to load assembly using a path not ending with ".dll" 
![image](https://github.com/dotnet/aspnetcore/assets/3961279/59b69230-2e23-4553-8e7c-85d54bee324c)

@davidfowl , @halter73 , fyi

**Comment by skoshelev** [comment]
@[batkaevruslan](https://github.com/batkaevruslan) Hi from nopCommerce team! Unfortunately, we cannot reproduce the problem locally.
However, while solving this problem for nopCommerce users, we encountered a situation that indirectly confirms the problem of finding a library without an extension. We got the full assembly name from the disk using the following code: `var an = AssemblyName.GetAssemblyName(dllPath);`. In this case, the assembly name is loaded without problems, since the dllPath corresponds to the real path to the library.
Then we tried to load the assembly itself using the following code: `var dll = AppDomain.CurrentDomain.Load(an);`. In this case we got the exception System.IO.FileNotFoundException
I think this is a pretty convincing confirmation of what you found when debugging your code.
